gprofiler/merge.py

```diff
@@ -82,7 +87,86 @@ def collapse_stack(stack: str, comm: str) -> str:
+    fp_perf, fp_pid_to_name = parse_perf_script(raw_fp_perf)
+    dwarf_perf, dwarf_pid_to_name = parse_perf_script(raw_dwarf_perf)
+    dwarf_pid_to_name.update(fp_pid_to_name)
+    merged_pid_to_name = dwarf_pid_to_name
+
+    if raw_fp_perf is None:
+        return dwarf_perf, merged_pid_to_name
+    elif raw_dwarf_perf is None:
+        return fp_perf, merged_pid_to_name
+
+    total_fp_samples = sum([sum(stacks.values()) for stacks in fp_perf.values()])
+    total_dwarf_samples = sum([sum(stacks.values()) for stacks in dwarf_perf.values()])
+    fp_to_dwarf_sample_ratio = total_fp_samples / total_dwarf_samples
+
+    # The FP perf is used here as the "main" perf, to which the DWARF perf is scaled.
+    merged_pid_to_stacks_counters: ProcessToStackSampleCounters = defaultdict(Counter)
+    add_highest_avg_depth_stacks_per_process(
+        dwarf_perf, fp_perf, fp_to_dwarf_sample_ratio, merged_pid_to_stacks_counters
+    )
+    total_merged_samples = sum([sum(stacks.values()) for stacks in merged_pid_to_stacks_counters.values()])
+    logger.debug(
+        f"Total FP samples: {total_fp_samples}; Total DWARF samples: {total_dwarf_samples}; "
+        f"FP to DWARF ratio: {fp_to_dwarf_sample_ratio}; Total merged samples: {total_merged_samples}"
+    )
+    return merged_pid_to_stacks_counters, merged_pid_to_name
+
+
+def add_highest_avg_depth_stacks_per_process(
+    dwarf_perf: ProcessToStackSampleCounters,
+    fp_perf: ProcessToStackSampleCounters,
+    fp_to_dwarf_sample_ratio: float,
+    merged_pid_to_stacks_counters: ProcessToStackSampleCounters,
+):
+    for pid, fp_collapsed_stacks_counters in fp_perf.items():
+        if pid not in dwarf_perf:
+            merged_pid_to_stacks_counters[pid] = fp_collapsed_stacks_counters
+            continue
+
+        fp_frame_count_average = get_average_frame_count(fp_collapsed_stacks_counters.keys())
+        dwarf_collapsed_stacks_counters = dwarf_perf[pid]
+        dwarf_frame_count_average = get_average_frame_count(dwarf_collapsed_stacks_counters.keys())
+        if fp_frame_count_average > dwarf_frame_count_average:
+            merged_pid_to_stacks_counters[pid] = fp_collapsed_stacks_counters
+        else:
+            dwarf_collapsed_stacks_counters = scale_dwarf_samples_count(
+                dwarf_collapsed_stacks_counters, fp_to_dwarf_sample_ratio
+            )
+            merged_pid_to_stacks_counters[pid] = dwarf_collapsed_stacks_counters
+
+
+def scale_dwarf_samples_count(
+    dwarf_collapsed_stacks_counters: StackToSampleCount, fp_to_dwarf_sample_ratio: float
+) -> StackToSampleCount:
+    if fp_to_dwarf_sample_ratio == 1:
+        return dwarf_collapsed_stacks_counters
+    # scale the dwarf stacks to the FP stacks to avoid skewing the results
+    for stack, sample_count in dwarf_collapsed_stacks_counters.items():
+        new_count = sample_count * fp_to_dwarf_sample_ratio
+        if new_count < 1 and random.random() > new_count:
```

Comment:
1. Explain the rational behind this randomness
2. Any reason we apply this logic only for the `< 1` case and not for all cases? If the `new_count` is `1.13` then I want 1 sample at 13% and 2 samples at 87%, no? Although, if the `new_count` is `1.13` then something is way off anyway lol